RT-258 Add Prometheus Metrics exports

joao-vieira-leao-reef opened the pull request:
**Intent:**
This is part of RT-258. The other part will soon be another pull request to the Django Template.

**Summary of changes:** 

We're exporting metrics to a customizable (through env vars) directory in the Prometheus .prom format. We have a default metric that states the last scan and push that occurred. Apart from it, there is already groundwork for making new metrics on the fly should we need to.

Commit:
Add Prometheus Metrics exports

We're exporting metrics to a customizable (through env vars) directory in the Prometheus .prom format.
We have a default metric that states the last scan and push that occurred. Apart from it, there is already groundwork for making new metrics on the fly should we need to.

emnoor-reef commented on src/vulnrelay/metrics/exporter.py:
```diff
@@ -0,0 +1,42 @@
+        return "\n".join([f"{name} {metric.last_output}" for name, metric in self.metrics.items()])
+
+    def check_dir(self) -> None:
+        dir = os.path.dirname(self.path)
+        if not os.path.exists(dir):
+            os.makedirs(dir)
```
nitpick: use `pathlib`, which would also avoid TOCTOU issue.

joao-vieira-leao-reef replied:
I made the switch, thanks for pointing this out.

emnoor-reef commented on src/vulnrelay/metrics/exporter.py:
```diff
@@ -0,0 +1,42 @@
+        path = self.path
+        temp_file = f"{path}.tmp"
+
+        if os.path.exists(path):
+            os.remove(path)
```
This is not necessary. `shutil.move` below should remove the old file. Also avoids TOCTOU issue.

emnoor-reef reviewed (approved):
The solution looks good. But IMO it is unnecessarily abstracted. A simple function that takes a path and a value (timestamp) and writes to the prom file would have been much better.

Unless there are other metrics planned to be exported, I don't see the value of abstraction from the get go.

Commit:
lessen abstractions for metrics and minor fixes

- removed metric class, as it was holding a single value
- switched to pathlib in exporter check_dir
- removed redundant file deletion call with shutil

joao-vieira-leao-reef commented:
> The solution looks good. But IMO it is unnecessarily abstracted. A simple function that takes a path and a value (timestamp) and writes to the prom file would have been much better.
> 
> Unless there are other metrics planned to be exported, I don't see the value of abstraction from the get go.

I agree in retrospect. I still kept *some* abstractions though, I believe this can benefit from it as more metrics are added should the demand arise. Still, I deleted the Metric class, since it was essentially holding a single value haha.

olzhasar-reef commented on src/vulnrelay/metrics/exporter.py:
```diff
@@ -0,0 +1,37 @@
+from typing import Any
+
+
+class MetricExporter:
```
Continuing on @emnoor-reef 's point:
Abstractions are considered good and helpful when they abstract the complexity away from the user, thus making life easier and / or reducing mental overhead.
This particular class requires the user to instantiate a dictionary with null (or non-null?) initial values. Both the data structure and initial values are essentially the implementation detail which final user should likely not be bothered with.

olzhasar-reef commented on src/vulnrelay/metrics/exporter.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class MetricExporter:
+    def __init__(self, *, metrics: dict[str, Any], path: str) -> None:
```
If we have a dedicated enum to restrict possible metrics values, it's unclear why don't we use it as a type.

And why `str` for the path argument and not `Pathlib.Path`?

joao-vieira-leao-reef replied:
This would improve consistency, yes. I'll work on that.

olzhasar-reef commented on src/vulnrelay/metrics/exporter.py:
```diff
@@ -0,0 +1,37 @@
+    def export_all(self) -> None:
+        content = self._get_export_content()
+        path = self.path
+        temp_file = f"{path}.tmp"
```
What problem are solving with having a temp file instead of writing directly to the final metrics file?

joao-vieira-leao-reef replied:
We write "atomically" to the file to avoid a known issue in node exporter where it could read a file halfway through the writing process

https://www.robustperception.io/atomic-writes-and-the-textfile-collector/

olzhasar-reef replied:
Even though in our case we'll likely always use a single system call (which should be atomic by itself), that's still a great touch!

olzhasar-reef commented on src/vulnrelay/services.py:
```diff
@@ -97,3 +111,10 @@ def run_workflow(
         logger.info("Upload successful")
+
+    try:
+        metric_exporter.save_and_export(
```
It would be nice to have some tests, to ensure that metrics are indeed being written and that possible failures do not hamper our core functionality.

olzhasar-reef commented on src/vulnrelay/metrics/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from .exporter import MetricExporter
+from .metric import MetricNames
+
+__all__ = ["MetricExporter", "Metric", "MetricNames"]
```
The `Metric` abstraction has been removed, I assume

olzhasar-reef commented on src/vulnrelay/services.py:
```diff
@@ -97,3 +111,10 @@ def run_workflow(
+
+    try:
+        metric_exporter.save_and_export(
+            values={MetricNames.LAST_SCAN_AND_PUSH.value: int(time.time())},
```
Prometheus docs seem to require milliseconds from epoch:
https://prometheus.io/docs/instrumenting/exposition_formats/
But `time.time()` will give us seconds instead?

joao-vieira-leao-reef replied:
This didn't show up as an issue in manual testing, but yeah I'll adjust

joaovl5 committed:
added test and fixed consistency issues

- removed redudant dict argument to metric exporter
- switched to pathlib on more instances
- added metric exporter test

joao-vieira-leao-reef commented:
made the requested changes (or I hope)

also pushed from wrong account sorry

olzhasar-reef commented on tests/test_exporter.py:
```diff
@@ -0,0 +1,16 @@
+from vulnrelay.metrics.metric import MetricNames
+
+
+def test_export_metrics():
```
This doesn't test much without actually verifying the output file contents

joao-vieira-leao-reef replied:
Done

joaovl5 committed:
improved test cases for exporter

joao-vieira-leao-reef commented:
Improved test case for exporter
